### PR TITLE
cam6_3_047: Move WACCM-x code in dp_coupling to common location

### LINKED ISSUE #444: Move WACCM-x code in dp_coupling to common module/subroutine for use by all dycores

### LINKED ISSUE #517: Double call of outfld for HNO3_GAS in mo_gas_phase_chemdr.F90
As soon as `has_strato_chem` is true, the following line is called twice in `src/chemistry/mozart/mo_gas_phase_chemdr.F90`:

`call outfld( 'HNO3_GAS',   vmr(:ncol,:,hno3_ndx), ncol, lchnk )`

This means that the sum of `HNO3_GAS`, `HNO3_STS` and `HNO3_NAT` is no more equal to `HNO3_TOTAL` in the output. Suggestion is to remove the second call of the above subroutine.

@fvitt

### PR DESCRIPTION
**Pull request by fvitt**
Duplicate WACCM-X specific code in dp_coupling for the FV, SE, and MPAS dycores is moved to a central common location (in physics_type module).

closes #444 

Fixes issue with HNO3_GAS diagnostic and minor fixes to history long names in chemistry

closes #517
closes #518

### CODE REVIEW

File: src/dynamics/se/dp_coupling.F90 [hunk 1]
```diff
@@ -560,14 +559,6 @@ subroutine derived_phys_dry(phys_state, phys_tend, pbuf2d)
...
 
    integer :: m, i, k, ncol
-
-   !--------------------------------------------
-   !  Variables needed for WACCM-X
-   !--------------------------------------------
-    real(r8) :: mmrSum_O_O2_H                ! Sum of mass mixing ratios for O, O2, and H
-    real(r8), parameter :: mmrMin=1.e-20_r8  ! lower limit of o2, o, and h mixing ratios
-    real(r8), parameter :: N2mmrMin=1.e-6_r8 ! lower limit of o2, o, and h mixing ratios
-
```
Comment (cacraigucar) [inline src/dynamics/se/dp_coupling.F90:570]: These variables need to be removed from src/dynamics/mpas/dp_coupling.F90 as well.  Can also remove H2lim in that module.

**Review by cacraigucar — APPROVED** [review]

File: src/dynamics/se/dp_coupling.F90 (hunk 1 above)
Comment (fvitt) [reply src/dynamics/se/dp_coupling.F90:570]: Good catch.  I have cleaned these out of the mpas version.

File: src/physics/cam/physics_types.F90 [hunk 2]
```diff
@@ -1137,6 +1137,59 @@ subroutine init_geo_unique(phys_state,ncol)
...
+
+       ncol = state%ncol
+
+       !------------------------------------------------------------
+       ! Ensure N2 = 1-(O2 + O + H) mmr is greater than 0
+       ! Check for unusually large H2 values and set to lower value.
+       !------------------------------------------------------------
+
+       do i=1,ncol
+          do k=1,pver
```
Comment (nusbaume) [inline src/physics/cam/physics_types.F90:1166]: Would it be slightly more efficient to flip the order of these loops (for column-major loop ordering)?

**Review by nusbaume — APPROVED** [review]
Looks good to me!  Just one minor request/question.

File: src/physics/cam/physics_types.F90 (hunk 2 above)
Comment (fvitt) [reply src/physics/cam/physics_types.F90]: done

**Review by gold2718 — APPROVED** [review]
One question and the ChangeLog needs finishing (obviously) but it looks good to me.

Thanks for doing this cleanup!

File: src/physics/cam/physics_types.F90 [hunk 3]
```diff
@@ -5,7 +5,7 @@ module physics_types
 
   use shr_kind_mod,     only: r8 => shr_kind_r8
   use ppgrid,           only: pcols, pver
-  use constituents,     only: pcnst, qmin, cnst_name
+  use constituents,     only: pcnst, qmin, cnst_name, cnst_get_ind
```
[the remark this answers is not in the document]
Comment (fvitt) [reply src/physics/cam/physics_types.F90:8]: I suppose I simply appended `cnst_get_ind` to the line.  Would there be an advantage to having the use statement at the subroutine level?

[the remark this answers is not in the document]
Comment (gold2718) [reply src/physics/cam/physics_types.F90:8]: It is not a big deal. In general, it makes code more modular and helps keep our imports fairly clean. Not important enough to fix, more something to keep in mind.

**Review by peverwhee — APPROVED** [review]

**Review by PeterHjortLauritzen — APPROVED** [review]

**Pull request merged**